In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

_______________________________________________________________________________________________________________________________

# LINK TO CHATGPT CONVERSATION IN VERY LAST CELL

_______________________________________________________________________________________________________________________________

In [3]:
df

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

In [2]:
import pandas as pd

# Load the dataset from the URL
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_data = pd.read_csv(url)

# Get the shape of the dataset (rows, columns)
rows, columns = villagers_data.shape

# Print the number of rows and columns
print(f"The dataset has {rows} rows and {columns} columns.")

The dataset has 391 rows and 11 columns.


# Question 2.2

### observation 
each row represents an observation of a data point, in my case of using villagers you can read from left to right on each row to find specific unique characterization of each villager

### Variable
the variables are represented in the columns in the dataset, each column's header in this case give different characterization to villagers (gender, birthday, personality, etc...)


In [6]:
import pandas as pd

# Load the dataset from the URL
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_data = pd.read_csv(url)

# Show the first 5 rows of the dataset
print("First 5 Rows of the Dataset:")
print("-" * 50)
print(villagers_data.head())
print("\n")

# Show summary information about the dataset (like number of columns, types, etc.)
print("Dataset Information:")
print("-" * 50)
villagers_data.info()
print("\n")

# Display summary statistics for numerical columns
print("Summary Statistics for Numerical Columns:")
print("-" * 50)
print(villagers_data.describe())
print("\n")

# Display value counts for each categorical column
print("Value Counts for Categorical Columns:")
print("=" * 50)
for column in villagers_data.columns:
    print(f"\nColumn: {column}")
    print("-" * 50)
    print(villagers_data[column].value_counts())
    print("=" * 50)
    print("\n")


First 5 Rows of the Dataset:
--------------------------------------------------
   row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3 

# QUESTION 4

In [13]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df1 = pd.read_csv(url)

# Sum of missing values per column
missing_per_column = df1.isna().sum()

# Sum of missing values per row
missing_per_row = df1.isna().sum(axis=1)

# Total number of missing values in the DataFrame
total_missing = missing_per_column.sum()
print(f"Total missing values in the DataFrame: {total_missing}")

# Optionally, you can also print the missing values per row
print("Missing values per row:")
print(missing_per_row)

Total missing values in the DataFrame: 869
Missing values per row:
0      1
1      0
2      1
3      0
4      1
      ..
886    1
887    0
888    2
889    0
890    1
Length: 891, dtype: int64


In [8]:
import pandas as pd
url = " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"  # using different csv file with missing values
df1 = pd.read_csv(url) 
df1.shape


(891, 15)

In [9]:
import pandas as pd
url = " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"  # using different csv file with missing values
df1 = pd.read_csv(url) 
df1.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

## Question 4 Answer
this question is about how df1.shape and df1.describe() deals with displaying or calculating missing values.
So first of all, df1.shape will give the dimensions of the rows and columns regardless of the type of data, which is good!
Secondly, df1.describe() will calculate only numerical values given and will NOT include missing values unless told otherwise which is a discrepancy.

Now, the count row represents the number of non-missing values reported by df1.describe() unless told to otherwise which stated before is a discrepancy since it is not an 100% clear accurate representation of the data


## Question 5 DIFFERENCE IN METHOD OR ATTRIBUTE
a method does an action of computation and will require parentheses. Example df.describe()

an attribute provides information about the data about the object and can be used without parentheses. Example df.shape

## Question 6 definition of each analytic in df.describe() 
#### count: 
the number of non-missing values in the column
#### mean: 
sum of all the values divided by the count of non-null values, so it it just the average of all values in the column
#### standard deviation: 
it states the deviation of values from the mean. the higher the number calculated the more variability.
#### min:
the min represents the lowest boundary of the data (lowest number in the dataset)
#### 25%:
 also known as the first quartile, 25% of the data falls below this mark.
#### 50%:
The median of the data, at this point 50% of the data is below and above this mark
#### 75%:
the third quartile divides the 75% from the 25% and now 75% of the data falls below this mark
#### Max:
This represent the highest possible value provided in the data


# Question 7.1
Use case: Cleaning a Dataset with Missing Values in Specific Columns

df.dropna() is more useful when you want to drop rows that are missing critical datasets or keep rows missing values of non-critical data. (Also You can use the subset to target specific columns to check for missing values)

However, in the case of del df['col'] it will delete an entire column without the same consideration to the data that df.dropna() could give which would not be helpful in our use case of cleaning the dataset

# Question 7.2
Use case: removing an unnecessary column

del df['col'] deletes a specified column from the DataFrame, this can be used to delete entire columns that are considered relevant to what you want to analyze.

in this case df.dropna() would not be useful because we want to delete a column and not handle missing values 

# Question 7.3
using del df['col'] before df.dropna() could be useful in the case of showing better clarity. By cleaning up the DataFrame and dropping column that are no longer useful to what you, you make sure that when you use df.dropna() you are only including relevant data. Which can make your process more easier to understand.

# Question 7.4
## Before
i'm going to use the titanic DataSet and will first display the before report of the data below

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df2 = pd.read_csv(url)

# Check the initial shape and missing data
print("Before cleaning:")
print(df2.shape)
print(df2.isna().sum())

Before cleaning:
(891, 15)
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


### Approach
i'm goin to remove 'deck' since it is the one with the most missing values using  del df2['col'] and im going to remove rows with missing values with df2.dropna()

In [6]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df2 = pd.read_csv(url)

del df2['deck']

# Drop rows with missing values in remaining columns
df2_cleaned = df2.dropna()

# Check the shape and missing data after cleaning
print("\nAfter cleaning:")
print(df2_cleaned.shape)
print(df2_cleaned.isna().sum())


After cleaning:
(712, 14)
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64


### Approach justification
deck had the most missing values meaning it had very little value in our data so removing it was necessary, then we used dropna() to remove rows with missing information in important columns.
### Before: 
dataset had 891 rows and 15 columns with missing values in important columns.
### After: 
cleaned dataset now has 712 rows and 14 columns with all missing data removed.

## Question 8.1 explaining df.groupby("col1")["col2"].describe()
df.groupby("col1") makes different groups of each unique value in col1. 

Then ["col2"].describe() this will make summary statistics for col2 in each group.
## EXAMPLE

In [7]:
import pandas as pd

url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

df.groupby("species")["personality"].describe().head()  # only displaying top 5, too many rows


count unique     top freq
species                            
alligator     7      5    lazy    2
anteater      7      6   peppy    2
bear         15      7  cranky    5
bird         13      7    jock    4
bull          6      3  cranky    3

## Question 8.2 DIFFERENCE  between  df.describe()   &    df.groupby("col1")["col2"].describe()
 df.describe() checks independently each column and does not pay attention to the relationship between them like  df.groupby("col1")["col2"].describe() does.
 
 df.groupby("col1")["col2"].describe() also gives a more descriptive breakdown. the count calculated is for col2 within each group of col1 and like df.describe() it only counts non-missing values.



_______________________________________________________________________________________________________________________________

# 8.3 A not importing pandas

In [1]:
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

NameError: name 'pd' is not defined

# 8.3A
### Chatgpt 
correctly told me to import pandas as pd in the first line of my code
### Google
a google search told me to import pandas before using it, but didn't tell me how and it took way more time than chatgpt

# 8.3B TYPO titanic => titanics

In [6]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)
df.isna().sum()

HTTPError: HTTP Error 404: Not Found

### chatgpt
I gave my code to chatgpt and the error code I got it, surprisingly told me that there is a typo in the url and provided the correct link instantly, very impressive.
### google
after a minute of looking i found that this error is related to invalid http link, but provided no further assistance like chatgpt could

## 8.3C Trying to use a dataframe before it's been assigned into the variable

In [7]:
import pandas as pd

url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

Df.groupby("species")["personality"].describe().head() 

NameError: name 'Df' is not defined

### Chatgpt
fixed my code and provided code to copy paste to fix my problem

### Google
google told me how to fix my problem but did not provide specific code like chatgpt did

## 8.4D Forget one parentheses somewhere

In [8]:
import pandas as pd

url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url
df.isna().sum()

SyntaxError: '(' was never closed (4205515142.py, line 4)

### chatgpt
chatgpt gave me code that added the missing parentheses and correctly identified the problem
### google
google correctly told me that i should carefully scan my code and look for any missing parenthesis

## 8.4E Mistype one of the names of the chained functions with the code

In [9]:
import pandas as pd

url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

df.group_by("col1")["col2"].describe()  


AttributeError: 'DataFrame' object has no attribute 'group_by'

### chatgpt 
chatgpt told me that i mistyped, instead of group_by i should have inputted grouby. not separated by an underscore
### google
google told me that I miss typed it but couldn't offer specific guidance like chatgpt did.

## 8.3F column name that's in data

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

df.groupby("species")["sex"].describe().head()

KeyError: 'Column not found: sex'

### chatgpt
chatgpt told me that 'sex' is not found in my dataframe and my best option would be to use real columns within my dataset
### google
google essentially told me the same thing that chatgpt did

## 8.3G Forget to put the column name as a string

In [2]:
import pandas as pd

url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

df.groupby(species)["personality"].describe().head()

NameError: name 'species' is not defined

### chatgpt
chatgpt gave me a code to copy and paste to fix my issue, and correctly identified where the problem was for me

### google 
i typed the error code into google and was told what might be causing this problem, not as useful as chatgpt.

# 8.3 SUMMARY
google was most definitely way slower than chatgpt and had far more limitations, for example what i would do is i would just copy paste my whole code into chatgpt with my error code and every time it would tell me precisely what my error was and how to fix it. Chatgpt also gave me code to copy and paste to fix my code! Chatgpt is definitely more efficient and precise when it comes to fixing bugs in code than google.


## Question 9 Have you reviewed the course wiki-textbook and interacted with a ChatBot
Yes

## CHATGPT LINK
first half of the homework https://chatgpt.com/share/fa4b951d-780a-446f-a745-b6d9bdf7cf1f

second half of the homework https://chatgpt.com/share/c78a428e-7656-42f4-b529-8679d1c3dd7d